src/loader/summary.rs

```diff
@@ -0,0 +1,726 @@
+
+            // TODO: How do we deal with headings like "# My **awesome** summary"?
+            // for now, I'm just going to scan through and concatenate the
+            // Event::Text tags, skipping any styling.
+            Some(stringify_events(tags))
+        } else {
+            None
+        }
+    }
+
+    /// Push a new section at the end of the current nesting level.
+    fn push_numbered_section(&mut self, item: SummaryItem) -> SectionNumber {
+        if let State::NumberedChapters(level) = self.state {
+            push_item_at_nesting_level(
+                &mut self.summary.numbered_chapters,
+                item,
+                level as usize,
+                SectionNumber::default(),
+            ).chain_err(|| {
+                format!("The parser should always ensure we add the next \
+                item at the correct level ({}:{})", module_path!(), line!())
+            })
+                .unwrap()
+        } else {
+            // this method should only ever be called when parsing a numbered
+            // section, therefore if we ever get here something has gone
+            // hideously wrong...
+            error!("Calling push_numbered_section() when not in a numbered section");
+            error!("Current state: {:?}", self.state);
+            error!("Item: {:?}", item);
+            error!("Summary:");
+            error!("{:#?}", self.summary);
+            panic!("Entered unreachable code, this is a bug");
+        }
+    }
+}
+
+/// Given a particular level (e.g. 3), go that many levels down the `Link`'s
+/// nested items then append the provided item to the last `Link` in the
+/// list.
+fn push_item_at_nesting_level(links: &mut Vec<SummaryItem>, mut item: SummaryItem, level: usize, mut section_number: SectionNumber)
+    -> Result<SectionNumber> {
+    if level == 0 {
+        // set the section number, if applicable
+        section_number.push(links.len() as u32 + 1);
+
+        if let SummaryItem::Link(ref mut l) = item {
+            l.number = Some(section_number.clone());
+        }
+
+        links.push(item);
+        Ok(section_number)
+    } else {
+        let next_level = level - 1;
+        let index_for_item = links.len() + 1;
+
+        // FIXME: This bit needs simplifying!
+        let (index, last_link) = get_last_link(links).chain_err(|| {
+            format!("The list of links needs to be {} levels deeper (current position {})", 
+                level, section_number)
+        })?;
+
+        section_number.push(index as u32 + 1);
+        push_item_at_nesting_level(&mut last_link.nested_items, item, level - 1, section_number)
+    }
+}
+
+/// Gets a pointer to the last `Link` in a list of `SummaryItem`s, and its
+/// index.
+fn get_last_link(links: &mut [SummaryItem]) -> Result<(usize, &mut Link)> {
+    links
+        .iter_mut()
+        .enumerate()
+        .filter_map(|(i, item)| item.maybe_link_mut().map(|l| (i, l)))
+        .rev()
```

Comment:
That's what I originally tried too. What you are trying to do is get the last chapter from the `sub_items`, but what happens when you call `last()` and the last item is a `Separator` instead of a `Chapter`? 

I agree that it's not overly pretty or easy to understand, I'll see if I can pull it out into a method.